client/src/app/core/ui-services/tree.service.ts

```diff
@@ -56,6 +77,169 @@ export class TreeService {
+     * @returns an array that contains the parent inclusive its children.
+     */
+    private buildTreeHelper<T extends Identifiable & Displayable>(
+        level: number,
+        parent: T,
+        mapChildren: Map<Number, T[]>
+    ): OSFlatNode[] {
+        const children = mapChildren.get(parent.id);
+
+        // build the parent node
+        const tmpArray: OSFlatNode[] = [
+            {
+                id: parent.id,
+                name: parent.getTitle(),
+                level: level,
+                expandable: !!children,
+                isExpanded: !!children
+            }
+        ];
+
+        // if there are children, check if they have children, too
+        if (children) {
+            for (let i = 0; i < children.length; ++i) {
+                const value = children[i];
+                tmpArray.push(...this.buildTreeHelper(level + 1, value, mapChildren));
+            }
+        }
+        return tmpArray;
+    }
+
+    /**
+     * Function, that returns a node containing information like id, name and children.
+     * Children only, if available.
+     *
+     * @param node The node which is converted.
+     * @param nodes The array with all nodes to convert.
+     * @param length The number of converted nodes related to the parent node.
+     *
+     * @returns `OSTreeNodeWithOutItem`
+     */
+    private getNodeWithChildren(
+        node: OSFlatNode,
+        nodes: OSFlatNode[],
+        parentId: number,
+        length: number
+    ): { tree: Array<OSTreeSortEvent>; length: number } {
+        const children = [];
+        // Begins at the position of the node in the array
+        // Ends if the next node has the same or higher level than the given node.
+        for (let i = node.position + 1; !!nodes[i] && nodes[i].level >= node.level + 1; ++i) {
+            const nextNode = nodes[i];
+            // The next node is a child if the level is one higher than the given node.
+            if (nextNode.level === node.level + 1) {
+                const child = this.getNodeWithChildren(nextNode, nodes, node.id, length);
+
+                length += child.length;
+                children.push(...child.tree);
+            }
+        }
+
+        const osNode: OSTreeNodeWithoutItem = {
+            id: node.id,
+            name: node.name
+        };
+
+        const result: { tree: OSTreeSortEvent[]; length: number } = {
+            tree: [
+                ...children,
+                {
+                    nodes: [osNode],
+                    parent_id: parentId
+                }
+            ],
+            length: ++length
+        };
+
+        return result;
+    }
+
+    /**
+     * Function to convert a flat tree to a nested tree built from `OSTreeNodeWithOutItem`.
+     *
+     * @param nodes The array of flat nodes, which should be converted.
+     *
+     * @returns The tree with nested information.
+     */
+    public getBasicTreeWithChildren(nodes: OSFlatNode[]): OSTreeSortEvent[] {
```

Comment:
`makeTreeFromFlatTree` to be consistent with naming. Use OSTreeNode<T>[] as a return type.